src/utils/signals/worker.ts

```diff
@@ -0,0 +1,100 @@
+import { HttpTransportType, HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
+import type { WorkerIncomingMessage, WorkerOutgoingMessage } from './types';
+import { Deferred, EventEmitter } from './utils';
+import { env } from '~/env/client.mjs';
+
+// --------------------------------
+// Types
+// --------------------------------
+interface SharedWorkerGlobalScope {
+  onconnect: (event: MessageEvent) => void;
+}
+const _self: SharedWorkerGlobalScope = self as any;
+
+let connection: HubConnection | undefined;
+const events: Record<string, (data: unknown) => void> = {};
+const deferred = new Deferred<void>();
+
+const emitter = new EventEmitter<{
+  connectionReady: undefined;
+  connectionClosed: { message?: string };
+  connectionError: { message?: string };
+  eventReceived: { target: string; payload: any };
+}>();
+
+const getConnection = async ({ token }: { token: string }) => {
+  if (connection) return connection;
+
+  connection = new HubConnectionBuilder()
+    .withUrl(`${env.NEXT_PUBLIC_SIGNALS_ENDPOINT}/hub`, {
+      accessTokenFactory: () => token,
+      skipNegotiation: true,
+      transport: HttpTransportType.WebSockets,
+    })
+    .withAutomaticReconnect()
+    .build();
+
+  await connection.start();
+  connection.onreconnected(() => emitter.emit('connectionReady', undefined));
```

Comment:
The only thing I'm not 100% sure of is whether this triggers on initial connection or not 🤔.  From what I can tell, that's not the case, so 2 things:

- 1: I suggest we wrap this into a try catch. Connection.start will throw an error if it fails to connect initially.
- 2: after awaiting `connection.start` you might wanna emit `connectionReady`. I see you have below on the `then` but I think it's fine placing it here unless I'm not seeing something